deprecate dtype getters from `torch.testing` namespace

pmeier opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63844
* __->__ #63554

Following https://github.com/pytorch/pytorch/pull/61840#issuecomment-884087809, this deprecates all the dtype getters publicly exposed in the `torch.testing` namespace. The reason for this twofold:

1. If someone is not familiar with the C++ dispatch macros PyTorch uses, the names are misleading. For example `torch.testing.floating_types()` will only give you `float32` and `float64` skipping `float16` and `bfloat16`.
2. The dtype getters provide very minimal functionality that can be easily emulated by downstream libraries.

We thought about [providing an replacement](https://gist.github.com/pmeier/3dfd2e105842ad0de4505068a1a0270a), but ultimately decided against it. The major problem is BC: by keeping it, either the namespace is getting messy again after a new dtype is added or we need to somehow version the return values of the getters.

Differential Revision: [D30662206](https://our.internmc.facebook.com/intern/diff/D30662206)

pmeier committed:
deprecate dtype getters from `torch.testing` namespace

[ghstack-poisoned]

pmeier commented on torch/testing/_core.py:
```diff
@@ -9,31 +9,13 @@
 from typing import cast, List, Optional, Tuple, Union
 import operator
 
+from ._dtype_getters import get_all_dtypes, get_all_complex_dtypes
+
```
We need these here, because `is_integral` uses them. This will not be a problem in the future, since `is_integral` is only used by `_compare_tensors_internal`, which in turn is only (indirectly) used by `torch.testing._internal.common_utils.TestCase.assertEqual`, which ultimately will be replaced by `torch.testing.assert_close` (#58981).

mruberry replied:
Would you add a brief comment with this explanation and a TODO to remove this?

pmeier commented on torch/testing/_deprecated.py:
```diff
@@ -84,3 +86,16 @@ def assert_allclose(
+        continue
+
+    fn = getattr(_dtype_getters, name)
+    globals()[name] = warn_deprecated(f"The unparametrized call can be replaced with {fn()}")(fn)
```
Should we include a reason as to why we deprecate / remove the dtype getters? Since there is no RFC for it, it could seem arbitrary.

pmeier committed:
Update on "deprecate dtype getters from `torch.testing` namespace"




[ghstack-poisoned]

pmeier committed:
Update on "deprecate dtype getters from `torch.testing` namespace"


Following https://github.com/pytorch/pytorch/pull/61840#issuecomment-884087809, this deprecates all the dtype getters publicly exposed in the `torch.testing` namespace. The reason for this twofold:

1. If someone is not familiar with the C++ dispatch macros PyTorch uses, the names are misleading. For example `torch.testing.floating_types()` will only give you `float32` and `float64` skipping `float16` and `bfloat16`.
2. The dtype getters provide very minimal functionality that can be easily emulated by downstream libraries.

We thought about [providing an replacement](https://gist.github.com/pmeier/3dfd2e105842ad0de4505068a1a0270a), but ultimately decided against it. The major problem is BC: by keeping it, either the namespace is getting messy again after a new dtype is added or we need to somehow version the return values of the getters.

[ghstack-poisoned]

pmeier committed:
Update on "deprecate dtype getters from `torch.testing` namespace"


Following https://github.com/pytorch/pytorch/pull/61840#issuecomment-884087809, this deprecates all the dtype getters publicly exposed in the `torch.testing` namespace. The reason for this twofold:

1. If someone is not familiar with the C++ dispatch macros PyTorch uses, the names are misleading. For example `torch.testing.floating_types()` will only give you `float32` and `float64` skipping `float16` and `bfloat16`.
2. The dtype getters provide very minimal functionality that can be easily emulated by downstream libraries.

We thought about [providing an replacement](https://gist.github.com/pmeier/3dfd2e105842ad0de4505068a1a0270a), but ultimately decided against it. The major problem is BC: by keeping it, either the namespace is getting messy again after a new dtype is added or we need to somehow version the return values of the getters.

[ghstack-poisoned]

pmeier committed:
Update on "deprecate dtype getters from `torch.testing` namespace"


Following https://github.com/pytorch/pytorch/pull/61840#issuecomment-884087809, this deprecates all the dtype getters publicly exposed in the `torch.testing` namespace. The reason for this twofold:

1. If someone is not familiar with the C++ dispatch macros PyTorch uses, the names are misleading. For example `torch.testing.floating_types()` will only give you `float32` and `float64` skipping `float16` and `bfloat16`.
2. The dtype getters provide very minimal functionality that can be easily emulated by downstream libraries.

We thought about [providing an replacement](https://gist.github.com/pmeier/3dfd2e105842ad0de4505068a1a0270a), but ultimately decided against it. The major problem is BC: by keeping it, either the namespace is getting messy again after a new dtype is added or we need to somehow version the return values of the getters.

[ghstack-poisoned]

mruberry commented on torch/testing/_internal/opinfo_helper.py:
```diff
@@ -4,21 +4,21 @@
-                           integral_types_and,
-                           all_types_and,
-                           )
+from torch.testing._internal.dtype_getters import (all_types_and_complex_and,
```
This can be more easily formatted:

```
import \
  ( ... )
```

without having to space all the imported elements so far to the right of the page

pmeier replied:
I try to refrain form pure formatting changes as much as possible, because they only create noise.

mruberry commented on torch/testing/_deprecated.py:
```diff
@@ -84,3 +86,21 @@ def assert_allclose(
+            f"For all other devices, it can be replaced with {fn('cpu')}."
+        )
+    else:
+        instructions = f"The unparametrized call can be replaced with {fn()}"
```
Let's include the names of the functions:

"The call to FOO can be replaced by enumerating the dtypes DTYPES"

mruberry commented on torch/testing/_deprecated.py:
```diff
@@ -84,3 +86,21 @@ def assert_allclose(
+
+    if name == "get_all_math_dtypes":
+        instructions = (
+            f"For CUDA devices, the call can be replaced with {fn('cuda')}. "
```
"For CUDA devices a call to get_all_math_dtypes can be replaced by enumerating the dtypes DTYPES"

?

pmeier replied:
Not sure about the "enumerating" though. The warning will show the container, e.g. `[torch.float32, torch.float64]`, so the call can actually be replaced by what is printed.

mruberry replied:
OK

mruberry commented on torch/testing/_deprecated.py:
```diff
@@ -84,3 +86,21 @@ def assert_allclose(
     )
+
+
+for name in _dtype_getters.__all__:
```
Add a comment here

mruberry commented on torch/testing/_internal/dtype_getters.py:
```diff
@@ -0,0 +1,4 @@
+"""The content of torch/testing/_dtype_getters.py should be moved here as soon as the deprecation period is over.
+"""
+
```
Add a comment here that (because of your clever handling of deprecation warnings above) these imports will now throw a deprecation notice.

Should we create a new file for dtype_getters or stuff them in common_utils.py?

pmeier replied:
I dislike large and unstructured files, so I would prefer an separate module. If you don't like the name, I mean there is no `common_dtype.py` yet, so this would be also an option.

pmeier replied:
> Add a comment here that (because of your clever handling of deprecation warnings above) these imports will now throw a deprecation notice.

The whole point of `torch/testing/_dtype_getters.py` is to have access to the getters without seeing a deprecation warning. So if you import our getters from there everything is as before. Only if you import them from `torch.testing` you will see the warning.

For now we could live without `torch/testing/_internal/dtype_getters.py`, but I wanted to avoid changing all imports again after the deprecation cycle is over and everything is internalized. By adding an internal module upfront, we only need to move the functionality there in the future, but all import stay the same.

mruberry replied:
Oh I meant that these WILL NOT THROW. Sorry, typing is hard.

mruberry replied:
Sure; this file seems fine.

mruberry commented on torch/testing/_dtype_getters.py:
```diff
@@ -0,0 +1,136 @@
+def all_types_and_half():
+    return _all_types_and_half
+
+def get_all_dtypes(include_half=True,
```
Let's add a comment separating the dtype functions based on dispatch macros from the test suites own "get" versions

mruberry commented:
Just a few inline comments for your review, @pmeier

pmeier committed:
Update on "deprecate dtype getters from `torch.testing` namespace"


Following https://github.com/pytorch/pytorch/pull/61840#issuecomment-884087809, this deprecates all the dtype getters publicly exposed in the `torch.testing` namespace. The reason for this twofold:

1. If someone is not familiar with the C++ dispatch macros PyTorch uses, the names are misleading. For example `torch.testing.floating_types()` will only give you `float32` and `float64` skipping `float16` and `bfloat16`.
2. The dtype getters provide very minimal functionality that can be easily emulated by downstream libraries.

We thought about [providing an replacement](https://gist.github.com/pmeier/3dfd2e105842ad0de4505068a1a0270a), but ultimately decided against it. The major problem is BC: by keeping it, either the namespace is getting messy again after a new dtype is added or we need to somehow version the return values of the getters.

[ghstack-poisoned]

pmeier committed:
Update on "deprecate dtype getters from `torch.testing` namespace"


Following https://github.com/pytorch/pytorch/pull/61840#issuecomment-884087809, this deprecates all the dtype getters publicly exposed in the `torch.testing` namespace. The reason for this twofold:

1. If someone is not familiar with the C++ dispatch macros PyTorch uses, the names are misleading. For example `torch.testing.floating_types()` will only give you `float32` and `float64` skipping `float16` and `bfloat16`.
2. The dtype getters provide very minimal functionality that can be easily emulated by downstream libraries.

We thought about [providing an replacement](https://gist.github.com/pmeier/3dfd2e105842ad0de4505068a1a0270a), but ultimately decided against it. The major problem is BC: by keeping it, either the namespace is getting messy again after a new dtype is added or we need to somehow version the return values of the getters.

[ghstack-poisoned]

pmeier committed:
Update on "deprecate dtype getters from `torch.testing` namespace"


Following https://github.com/pytorch/pytorch/pull/61840#issuecomment-884087809, this deprecates all the dtype getters publicly exposed in the `torch.testing` namespace. The reason for this twofold:

1. If someone is not familiar with the C++ dispatch macros PyTorch uses, the names are misleading. For example `torch.testing.floating_types()` will only give you `float32` and `float64` skipping `float16` and `bfloat16`.
2. The dtype getters provide very minimal functionality that can be easily emulated by downstream libraries.

We thought about [providing an replacement](https://gist.github.com/pmeier/3dfd2e105842ad0de4505068a1a0270a), but ultimately decided against it. The major problem is BC: by keeping it, either the namespace is getting messy again after a new dtype is added or we need to somehow version the return values of the getters.

[ghstack-poisoned]

pmeier committed:
Update on "deprecate dtype getters from `torch.testing` namespace"


Following https://github.com/pytorch/pytorch/pull/61840#issuecomment-884087809, this deprecates all the dtype getters publicly exposed in the `torch.testing` namespace. The reason for this twofold:

1. If someone is not familiar with the C++ dispatch macros PyTorch uses, the names are misleading. For example `torch.testing.floating_types()` will only give you `float32` and `float64` skipping `float16` and `bfloat16`.
2. The dtype getters provide very minimal functionality that can be easily emulated by downstream libraries.

We thought about [providing an replacement](https://gist.github.com/pmeier/3dfd2e105842ad0de4505068a1a0270a), but ultimately decided against it. The major problem is BC: by keeping it, either the namespace is getting messy again after a new dtype is added or we need to somehow version the return values of the getters.

[ghstack-poisoned]

mruberry reviewed (approved):
Updates look good. Let's wait until after the 1.10 branch cut for safety and we'll get this landed

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30662206).

mruberry commented:
Hey @pmeier -- bad news is that the internal system is still struggling to merge this. Would you try rebasing this PR and pinging me and then I can reimport it and insta-merge it?

pmeier committed:
Update on "deprecate dtype getters from `torch.testing` namespace"


Following https://github.com/pytorch/pytorch/pull/61840#issuecomment-884087809, this deprecates all the dtype getters publicly exposed in the `torch.testing` namespace. The reason for this twofold:

1. If someone is not familiar with the C++ dispatch macros PyTorch uses, the names are misleading. For example `torch.testing.floating_types()` will only give you `float32` and `float64` skipping `float16` and `bfloat16`.
2. The dtype getters provide very minimal functionality that can be easily emulated by downstream libraries.

We thought about [providing an replacement](https://gist.github.com/pmeier/3dfd2e105842ad0de4505068a1a0270a), but ultimately decided against it. The major problem is BC: by keeping it, either the namespace is getting messy again after a new dtype is added or we need to somehow version the return values of the getters.

Differential Revision: [D30662206](https://our.internmc.facebook.com/intern/diff/D30662206)

[ghstack-poisoned]

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30662206).

pmeier committed:
Update on "deprecate dtype getters from `torch.testing` namespace"


Following https://github.com/pytorch/pytorch/pull/61840#issuecomment-884087809, this deprecates all the dtype getters publicly exposed in the `torch.testing` namespace. The reason for this twofold:

1. If someone is not familiar with the C++ dispatch macros PyTorch uses, the names are misleading. For example `torch.testing.floating_types()` will only give you `float32` and `float64` skipping `float16` and `bfloat16`.
2. The dtype getters provide very minimal functionality that can be easily emulated by downstream libraries.

We thought about [providing an replacement](https://gist.github.com/pmeier/3dfd2e105842ad0de4505068a1a0270a), but ultimately decided against it. The major problem is BC: by keeping it, either the namespace is getting messy again after a new dtype is added or we need to somehow version the return values of the getters.

Differential Revision: [D30662206](https://our.internmc.facebook.com/intern/diff/D30662206)

[ghstack-poisoned]

pmeier committed:
Update on "deprecate dtype getters from `torch.testing` namespace"


Following https://github.com/pytorch/pytorch/pull/61840#issuecomment-884087809, this deprecates all the dtype getters publicly exposed in the `torch.testing` namespace. The reason for this twofold:

1. If someone is not familiar with the C++ dispatch macros PyTorch uses, the names are misleading. For example `torch.testing.floating_types()` will only give you `float32` and `float64` skipping `float16` and `bfloat16`.
2. The dtype getters provide very minimal functionality that can be easily emulated by downstream libraries.

We thought about [providing an replacement](https://gist.github.com/pmeier/3dfd2e105842ad0de4505068a1a0270a), but ultimately decided against it. The major problem is BC: by keeping it, either the namespace is getting messy again after a new dtype is added or we need to somehow version the return values of the getters.

Differential Revision: [D30662206](https://our.internmc.facebook.com/intern/diff/D30662206)

[ghstack-poisoned]

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30662206).
